README.md
# kitchen-miasma

Provision Test Kitchen instances via [miasma](https://github.com/miasma-rb/miasma).

Inspired by [kitchen-ec2](https://github.com/test-kitchen/kitchen-ec2) and [kitchen-ssh](https://github.com/neillturner/kitchen-ssh).

This driver is experimental and should not be considered stable or suitable for any other use.


CHANGELOG.md
# kitchen-miasma changelog

## 0.0.1
- Initial implementation


spec/miasma_spec.rb
require 'logger'
require 'stringio'

require 'pry'

require 'kitchen/driver/miasma'
require 'kitchen/provisioner/dummy'
require 'kitchen/transport/dummy'
require 'kitchen/verifier/dummy'

describe Kitchen::Driver::Miasma do
  let(:logged_output) { StringIO.new }
  let(:logger)        { Logger.new(logged_output) }
  let(:config) { { :kitchen_root => '/kroot', :key_name => 'provider_key' } }
  let(:platform) { Kitchen::Platform.new(:name => 'ubuntu-14.04') }
  let(:suite) { Kitchen::Suite.new(:name => 'default') }
  let(:verifier) { Kitchen::Verifier::Dummy.new }
  let(:provisioner) { Kitchen::Provisioner::Dummy.new }
  let(:transport) { Kitchen::Transport::Dummy.new }
  let(:state_file) { double('state_file') }
  let(:state) { Hash.new }

  let(:non_default_compute_provider) do
    {
      :name => 'open_stack',
      :open_stack_username => 'alice',
      :open_stack_password => 'secret',
      :open_stack_region => 'us-west-2'
    }
  end

  let(:env) do
    {
      'AWS_DEFAULT_REGION' => 'us-west-2',
      'AWS_ACCESS_KEY_ID' => 'INVALID_ID',
      'AWS_SECRET_ACCESS_KEY' => 'INVALID_KEY'
    }
  end

  let(:default_data_fixture) do
    Smash.new(
      'default_flavor_id' => 'm3.medium',
      'usernames' => {
        'ubuntu-14.04' => 'ubuntu'
      },
      'regions' => {
        'us-west-2' => {
          'ubuntu-14.04' => 'ami-159cad25'
        }
      }
    )
  end

  let(:non_default_data_fixture) do
    Smash.new(
      'default_flavor_id' => 'm1.small',
      'usernames' => {
        'ubuntu-14.04' => 'openstack'
      },
      'regions' => {
        'us-west-2' => {
          'ubuntu-14.04' => 'ubuntu-image'
        }
      }
    )
  end

  let(:driver_object) { Kitchen::Driver::Miasma.new(config) }

  let(:driver) do
    d = driver_object
    instance
    d
  end

  let(:instance) do
    Kitchen::Instance.new(
      :verifier => verifier,
      :driver => driver_object,
      :logger => logger,
      :suite => suite,
      :platform => platform,
      :provisioner => provisioner,
      :transport => transport,
      :state_file => state_file
    )
  end

  before { stub_const("ENV", env) }

  it 'plugin_version is set to Kitchen::Driver::MIASMA_VERSION' do
    expect(driver.diagnose_plugin[:version]).to eq(Kitchen::Driver::MIASMA_VERSION)
  end

  context 'without a key path provided' do
    before do
      config[:key_name] = nil
    end

    it 'raises an exception' do
      expect { driver.verify_dependencies }.to raise_error(
        Kitchen::UserError, /key_name/
      )
    end
  end

  describe 'configuration' do

    it 'sets :key_path to nil by default' do
      expect(driver[:key_path]).to eq(nil)
    end

    it 'sets :username to nil by default' do
      expect(driver[:username]).to eq(nil)
    end

    it 'sets :port to 22 by default' do
      expect(driver[:port]).to eq(22)
    end

    it 'sets :retryable_tries to 60 by default' do
      expect(driver[:retryable_tries]).to eq(60)
    end

    it 'sets :retryable_sleep to 5 by default' do
      expect(driver[:retryable_sleep]).to eq(5)
    end

    it 'uses AWS credentials from environment variables as the defaults for :compute_provider' do
      expect(driver[:compute_provider]).to eq(
        :name => 'aws',
        :aws_region => 'us-west-2',
        :aws_access_key_id => 'INVALID_ID',
        :aws_secret_access_key => 'INVALID_KEY'
      )
    end

    context 'when the default :compute_provider is used' do
      before do
        allow_any_instance_of(Kitchen::Driver::Miasma).to receive(:images).and_return(default_data_fixture)
      end

      it 'returns the expected image id' do
        expect(driver[:image_id]).to eq('ami-159cad25')
      end
    end

    context 'when :comptue_provider is assigned a non-default value' do

      before do
        config[:compute_provider] = non_default_compute_provider
        allow_any_instance_of(Kitchen::Driver::Miasma).to receive(:images).and_return(non_default_data_fixture)
      end

      it 'overrides the default compute_provider' do
        expect(driver[:compute_provider]).to eq(non_default_compute_provider)
      end

      it 'returns the expected default flavor id' do
        expect(driver[:flavor_id]).to eq('m1.small')
      end

      it 'returns the expected default image id' do
        expect(driver[:image_id]).to eq('ubuntu-image')
      end

    end

  end

end



lib/kitchen/driver/miasma.rb
require 'json'
require 'miasma'
require 'pry'
require 'retryable'

require 'kitchen'
require 'kitchen/miasma/version'

module Kitchen
  module Driver
    class Miasma < SSHBase

      plugin_version Kitchen::Driver::MIASMA_VERSION

      default_config(:username, nil)
      default_config(:key_name, nil)
      default_config(:key_path, nil)
      default_config(:sudo, true)
      default_config(:port, 22)
      default_config(:retryable_tries, 60)
      default_config(:retryable_sleep, 5)
      default_config(:compute_provider) do
        {
          :name => 'aws',
          :aws_region => ENV['AWS_DEFAULT_REGION'],
          :aws_access_key_id => ENV['AWS_ACCESS_KEY_ID'],
          :aws_secret_access_key => ENV['AWS_SECRET_ACCESS_KEY']
        }
      end

      default_config(:image_id) do |driver|
        driver.default_image_id
      end

      default_config(:flavor_id) do |driver|
        driver.default_flavor_id
      end

      required_config(:key_name)

      def compute
        @compute ||= ::Miasma.api(
          :type => 'compute',
          :provider => config[:compute_provider][:name],
          :credentials => config[:compute_provider]
        )
      end

      # Returns defaults for known compute providers
      # @return [Hash]
      def images
        @images ||= begin
                      json_file = File.join(
                        File.dirname(__FILE__), %w(.. .. .. data), "#{config[:compute_provider][:name]}.json"
                      )
                      if File.exist?(json_file)
                        Hash.new(JSON.load(IO.read(json_file)))
                      else
                        warn("Failed to load defaults for #{config[:compute_provider][:name]} provider.")
                        Hash.new
                      end
                    end
      end

      # Returns the default image ID for the compute provider's given region.
      # @returns [String] default image ID
      def default_image_id
        region_name = config[:compute_provider]["#{config[:compute_provider][:name]}_region".to_sym]
        region_map = images.fetch('regions', {}).fetch(region_name, {})
        image_id = region_map && region_map[instance.platform.name]

        if image_id.nil?
          error("Could not determine default image_id in #{region_name} region for platform #{instance.platform.name}")
        end

        image_id
      end

      # Retrieve provider's default flavor id from image map
      # @return
      def default_flavor_id
        flavor_id = images['default_flavor_id']

        if flavor_id.nil?
          error("Could not determine default flavor_id for platform #{instance.platform.name} via #{config[:compute_provider][:name]}")
        end

        flavor_id
      end

      def configure_transport(state)
        if instance.transport[:username] == instance.transport.class.defaults[:username]
          image_username = images['usernames'][instance.platform.name]
          state[:username] = image_username if image_username
        end

        if config[:key_path]
          state[:ssh_key] = config[:key_path]
        end

        if config[:port]
          state[:port] = config[:port]
        end
      end

      def provision_server
        server = compute.servers.build(instance_data)
        server.save
        server
      rescue => e
        "Error provisioning server: #{e}"
      end

      def instance_data
        {
          :name => instance.name,
          :flavor_id => config[:flavor_id],
          :image_id => config[:image_id],
          :key_name => config[:key_name]
        }
      end

      def create(state)
        return if state[:server_id]

        configure_transport(state)

        server = provision_server
        state[:server_id] = server.id

        Retryable.retryable(
          :tries => config[:retryable_tries],
          :sleep => config[:retryable_sleep],
          :on => TimeoutError
        ) do |retries, _exception|

          c = retries * config[:retryable_sleep]
          t = config[:retryable_tries] * config[:retryable_sleep]
          info "Waited #{c}/#{t} for instance #{state[:server_id]} to become ready"
          server.reload
          ready = server.state == :running
          unless ready
            raise TimeoutError
          end
        end

        info("Instance #{state[:server_id]} created")
        state[:hostname] = server.address
        instance.transport.connection(state).wait_until_ready
      end

      def destroy(state)
        return unless state[:server_id]
        instance.transport.connection(state).close
        servers = compute.servers.all.select { |x| x.id == state[:server_id] }
        server = servers.first
        server.destroy

        Retryable.retryable(
          :tries => config[:retryable_tries],
          :sleep => config[:retryable_sleep],
          :on => TimeoutError
        ) do |retries, _exception|

          c = retries * config[:retryable_sleep]
          t = config[:retryable_tries] * config[:retryable_sleep]
          info "Waited #{c}/#{t} for instance #{state[:server_id]} to be destroyed"
          server.reload
          destroyed = server.state == :terminated
          unless destroyed
            raise TimeoutError
          end
        end

        state.delete(:server_id)
        state.delete(:hostname)
      end

    end

  end
end
